xline-client/src/clients/watch.rs

```diff
@@ -26,4 +31,268 @@ impl WatchClient {
+        let (mut request_sender, request_receiver) =
+            channel::<xlineapi::WatchRequest>(CHANNEL_SIZE);
+
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(request.into())),
+        };
+
+        request_sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))?;
+
+        let mut response_stream = self.inner.watch(request_receiver).await?.into_inner();
+
+        let watch_id = match response_stream.message().await? {
+            Some(resp) => {
+                assert!(resp.created, "not a create watch response");
+                resp.watch_id
+            }
+            None => {
+                return Err(ClientError::WatchError(String::from(
+                    "failed to create watch",
+                )));
+            }
+        };
+
+        Ok((Watcher::new(watch_id, request_sender), response_stream))
+    }
+}
+
+/// The watching handle.
+#[derive(Debug)]
+pub struct Watcher {
+    /// Id of the watcher
+    watch_id: i64,
+    /// The channel sender
+    sender: Sender<xlineapi::WatchRequest>,
+}
+
+impl Watcher {
+    /// Creates a new `Watcher`.
+    #[inline]
+    const fn new(watch_id: i64, sender: Sender<xlineapi::WatchRequest>) -> Self {
+        Self { watch_id, sender }
+    }
+
+    /// The ID of the watcher.
+    #[inline]
+    #[must_use]
+    pub const fn watch_id(&self) -> i64 {
+        self.watch_id
+    }
+
+    /// Watches for events happening or that have happened.
+    ///
+    /// # Errors
+    ///
+    /// If sender fails to send to channel
+    #[inline]
+    pub fn watch(&mut self, request: WatchRequest) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(request.into())),
+        };
+
+        self.sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))
+    }
+
+    /// Cancels this watcher.
+    ///
+    /// # Errors
+    ///
+    /// If sender fails to send to channel
+    #[inline]
+    pub fn cancel(&mut self) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CancelRequest(WatchCancelRequest {
+                watch_id: self.watch_id,
+            })),
+        };
+
+        self.sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))
+    }
+
+    /// Cancels watch by specified `watch_id`.
+    ///
+    /// # Errors
+    ///
+    /// If sender fails to send to channel
+    #[inline]
+    pub fn cancel_by_id(&mut self, watch_id: i64) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CancelRequest(WatchCancelRequest { watch_id })),
+        };
+
+        self.sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))
+    }
```

Comment:
Can `Watcher` cancel other `Watcher`s ?

Reply:
I think a `Watcher` is only responsible for the one watch request stream created by `WatchClient::watch`. So If two watcher is created they will not interfere with each other.